src/Modules/SimplCommerce.Module.Core/Controllers/CustomerGroupApiController.cs

```diff
@@ -0,0 +1,197 @@
+                }
+
+                if (search.CreatedOn != null)
+                {
+                    if (search.CreatedOn.before != null)
+                    {
+                        DateTimeOffset before = search.CreatedOn.before;
+                        query = query.Where(x => x.CreatedOn <= before);
+                    }
+
+                    if (search.CreatedOn.after != null)
+                    {
+                        DateTimeOffset after = search.CreatedOn.after;
+                        query = query.Where(x => x.CreatedOn >= after);
+                    }
+                }
+            }
+
+            var customerGroups = query.ToSmartTableResult(
+                param,
+                x => new
+                {
+                    Id = x.Id,
+                    Name = x.Name,
+                    Description = x.Description,
+                    IsActive = x.IsActive,
+                    CreatedOn = x.CreatedOn
+                });
+
+            return Json(customerGroups);
+        }
+        public IActionResult Get()
+        {
+            var customerGroups = _customergroupRepository.Query().Select(x => new
+            {
+                Id = x.Id,
+                Name = x.Name
+            });
+
+            return Json(customerGroups);
+        }
+
+        [HttpGet("{id}")]
+        public async Task<IActionResult> Get(long id)
+        {
+            var customergroup = await _customergroupRepository.Query().Include(x => x.Users).FirstOrDefaultAsync(x => x.Id == id);
+            var model = new CustomerGroupForm
+            {
+                Id = customergroup.Id,
+                Name = customergroup.Name,
+                Description = customergroup.Description,
+                IsActive = customergroup.IsActive
+            };
+
+            return Json(model);
+        }
+
+        [HttpPost]
+        public IActionResult Post([FromBody] CustomerGroupForm model)
+        {
+            try
+            {
+                if (ModelState.IsValid)
+                {
+                    var customergroup = new CustomerGroup
+                    {
+                        Name = model.Name,
+                        Description = model.Description,
+                        IsActive = model.IsActive
+                    };
+                    _customergroupRepository.Add(customergroup);
+                    _customergroupRepository.SaveChange();
+                    return Ok();
+                }
+            }
+            catch (DbUpdateException ex)
+            {
+                if (ExceptionHelper.IsUniqueConstraintViolation(ex))
+                {
+                    ModelState.AddModelError("Name", $"The Name '{model.Name}' is already in use, please enter a different name.");
+                    return new BadRequestObjectResult(ModelState);
+                }
+                ModelState.AddModelError("", "Unable to create customer group.  Please try again.");
```

Comment:
We will loose the exception information

Reply:
Yes, if it is not due to the unique constraint violation it will drop back to a generic message.  If that is what you mean?  

As discussed, I'll revise to check via code versus through exception handling.  Consequently,  risk of DB concurrency issues come into play.

Reply:
If another exception happen then all we get is "Unable to create customer group.  Please try again.". We don't know actually what happen inside.

Reply:
After a bit of thought, I would recommend we avoid query + insert versus exception handling. I think it would affect scalability - possibly performance (assumption), and would introduce race conditions / concurrency troubles.   

I think the exception handling can be revamped (new Issue?).  There's undoubtedly a few ways to attack it - globally via custom middleware > might < be one option.  Another is to modify SaveChange in Core.Data.RepositoryWithTypedId - wrapping the context.SaveChanges in try/catch and handle there (if we wanted to notify the user of the duplicate name - we would have to throw the exception back up the callstack and catch in the controller to set the ModelError message *I think*).  And there are probably a few other options that are not coming to mind.

Either way, whatever you prefer - we can go from there, let me know your thoughts.